// app_models/query_model.js
/**
 * User: osavch
 * Date: 28.10.13
 * Time: 10:42
 * File name:
 */

var mysql = require('mysql'),
    _ = require('underscore'),
    connection = mysql.createPool({
        host: 'localhost',
        user: 'root',
        password: '',
        database: 'hr_db',
        timezone: 'UTC+2'
    });


/**
 * Get positions
 * @param req
 * @param res
 */
exports.getPositions = function (req, res) {
    var sql = "SELECT " +
        "positions . *, " +
        "departments.name as department " +
        "FROM " +
        "positions " +
        "JOIN " +
        "departments " +
        "ON positions.department_id = departments.id " +
        "ORDER BY positions.id DESC "


    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}

/**
 * Insert postilion to DB
 * @param req
 * @param res
 */
exports.setPosition = function (req, res) {
    connection.query('INSERT INTO positions SET ?', req.body, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Delete Position
 * @param req
 * @param res
 */
exports.deletePosition = function (req, res) {
    var params = req.params,
        sql = "DELETE FROM positions WHERE id = ?",
        inserts = [params['id']];
    sql = mysql.format(sql, inserts);

    connection.query(sql, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Update position
 * @param req
 * @param res
 */
exports.updatePosition = function (req, res) {
    var data = req.body,
        id = data.id;
    delete  data.department;
    delete  data.id;

    connection.query('UPDATE positions SET ? where id = ' + id, data, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Get list of Departments
 * @param req
 * @param res
 */
exports.getDepartments = function (req, res) {
    var sql = "SELECT * FROM hr_db.departments ORDER BY departments.id DESC";
    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}

/**
 * Insert Department to DB
 * @param req
 * @param res
 */
exports.setDepartment = function (req, res) {
    connection.query('INSERT INTO departments SET ?', req.body, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}
/**
 * Delete Department
 * @param req
 * @param res
 */
exports.deleteDepartment = function (req, res) {
    var params = req.params,
        sql = "DELETE FROM departments WHERE id = ?",
        inserts = [params['id']];
    sql = mysql.format(sql, inserts);

    connection.query(sql, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Update Department
 * @param req
 * @param res
 */
exports.updateDepartment = function (req, res) {
    var data = req.body,
        id = data.id;
    delete  data.id;

    connection.query('UPDATE departments SET ? where id = ' + id, data, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Get list of Staff
 * @param req
 * @param res
 */
exports.getStaff = function (req, res) {
    var sql = "SELECT " +
        "staff.id, " +
        "staff.name, " +
        "staff.surname, " +
        "staff.middle_name, " +
        "staff.birthday, " +
        "staff.city, " +
        "staff.salary, " +
        "staff.startDate, " +
        "staff.date as modified, " +
        "positions.name as pasition, " +
        "positions.skills as skill, " +
        "departments.name as department " +
        "FROM " +
        "staff " +
        "JOIN " +
        "positions " +
        "JOIN " +
        "departments ON staff.department_id = departments.id " +
        "AND staff.position_id = positions.id " +
        "ORDER BY staff.id DESC "


    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}
/**
 * Insert Employee to DB
 * @param req
 * @param res
 */
exports.setEmployee = function (req, res) {
    connection.query('INSERT INTO staff SET ?', req.body, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}
/**
 * Delete Employee
 * @param req
 * @param res
 */
exports.deleteEmployee = function (req, res) {
    var params = req.params,
        sql = "DELETE FROM staff WHERE id = ?",
        inserts = [params['id']];
    sql = mysql.format(sql, inserts);

    connection.query(sql, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}
/**
 * Update Employee
 * @param req
 * @param res
 */
exports.updateEmployee = function (req, res) {
    var data = req.body,
        id = data.id;
    delete  data.id;

    connection.query('UPDATE staff SET ? where id = ' + id, data, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Returns two arrays
 *  Departments
 *  Positions
 * @param req
 * @param res
 */
exports.getAddEditEmployeeData = function (req, res) {
    var sql = "SELECT " +
        "departments.id as departmentId, " +
        "departments.name as departmentName, " +
        "positions.id as positionId, " +
        "positions.name as positionsName, " +
        "positions.skills as positionsSkill, " +
        "positions.department_id as posDepartId " +
        "FROM " +
        "departments " +
        "JOIN " +
        "positions " +
        "WHERE " +
        "positions.department_id = departments.id";


    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        var result = {
            positions: _collectPostions(rows),
            departments: _collectDepartments(rows)
        }
        res.setHeader('Content-Type', 'application/json');
        res.send(result);
    });
}

_collectPostions = function (rows) {
    var positions = [];
    _.each(rows, function (obj) {
        positions.push({
            positionId: obj.positionId,
            positionsName: obj.positionsName,
            positionsSkill: obj.positionsSkill,
            posDepartId: obj.posDepartId
        })
    });

    return _.sortBy(positions, function (obj) {
        return obj.positionsName
    });
}

_collectDepartments = function (rows) {
    var departments = [],
        departmentId = null;

    _.each(rows, function (obj) {
        if (departmentId === obj.departmentId) return;
        departmentId = obj.departmentId;
        departments.push({
            departmentId: obj.departmentId,
            departmentName: obj.departmentName
        })
    });

    return _.sortBy(departments, function (obj) {
        return obj.departmentName
    })
}

/**
 * Find user by Email
 * @param email
 * @param cb
 */
exports.findUser = function (email, cb) {
    var sql = "select " +
        "users . * " +
        "from " +
        "users " +
        "where " +
        "users.email = " + connection.escape(email);

    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;
        cb(err, rows);
    });
}

/**
 * Add new Admin or User
 * @param userData - fields array
 * @param cb
 */
exports.addNewUser = function (userData, cb) {
    connection.query('INSERT INTO users SET ?', userData, function (err, result) {
        if (err) throw err;
        cb(err, {success: true});
    });
}

/**
 * Get list of admins
 * @param req
 * @param res
 */
exports.getUsers = function (req, res) {
    var sql = "select " +
        "users.id, " +
        "users.name, " +
        "users.email, " +
        "users.position, " +
        "users.filial, " +
        "users.role, " +
        "users.date " +
        "from users " +
        "order by users.id desc ";

    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}

/**
 * Delete Manager or Admin
 * @param req
 * @param res
 */
exports.deleteManagerAdmin = function (req, res) {
    var params = req.params;
    if (parseInt(params['id']) === 1) {
        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    } else {

       var  sql = "DELETE FROM users WHERE id = ?",
            inserts = [params['id']];
        sql = mysql.format(sql, inserts);

        connection.query(sql, function (err, result) {
            if (err) throw err;

            res.setHeader('Content-Type', 'application/json');
            res.send({success: true});
        });
    }
}

/**
 * Update user record
 * @param req
 * @param res
 */
exports.updateUser = function (req, res, credentials) {
    var data = req.body,
        id = data.id;
    if(credentials){
        data.salt = credentials.salt;
        data.hash = credentials.hash;
    }
    delete data.id;
    delete data.password

    connection.query('UPDATE users SET ? where id = ' + id, data, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}
