// app_models/query_model.js
/**
 * User: osavch
 * Date: 28.10.13
 * Time: 10:42
 * File name:
 */

var mysql = require('mysql'),
    _ = require('underscore'),
    connection = mysql.createPool({
        host: 'localhost',
        user: 'root',
        password: '',
        database: 'hr_db',
        timezone: 'UTC+2'
    });


/**
 * Get positions
 * @param req
 * @param res
 */
exports.getPositions = function (req, res) {
    var sql = "SELECT " +
                "positions . *, " +
                "departments.name as department " +
               "FROM " +
                "positions " +
               "JOIN " +
                "departments " +
               "ON positions.department_id = departments.id " +
               "ORDER BY positions.id DESC "


    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}

/**
 * Insert postilion to DB
 * @param req
 * @param res
 */
exports.setPosition = function (req, res) {
    connection.query('INSERT INTO positions SET ?', req.body, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Delete Position
 * @param req
 * @param res
 */
exports.deletePosition = function (req, res) {
    var params = req.params,
        sql = "DELETE FROM positions WHERE id = ?",
        inserts = [params['id']];
    sql = mysql.format(sql, inserts);

    connection.query(sql, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Update position
 * @param req
 * @param res
 */
exports.updatePosition = function (req, res) {
    var data = req.body,
        id = data.id;
    delete  data.department;
    delete  data.id;

    connection.query('UPDATE positions SET ? where id = '+ id , data, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}

/**
 * Get list of Departments
 * @param req
 * @param res
 */
exports.getDepartments = function (req, res) {
    var sql = "SELECT * FROM hr_db.departments ORDER BY departments.id DESC";
    connection.query(sql, function (err, rows, fields) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send(rows);
    });
}

/**
 * Insert Department to DB
 * @param req
 * @param res
 */
exports.setDepartment = function (req, res) {
    connection.query('INSERT INTO departments SET ?', req.body, function (err, result) {
        if (err) throw err;

        res.setHeader('Content-Type', 'application/json');
        res.send({success: true});
    });
}
